### DISCUSSION TITLE
Assert function called by another function

### DISCUSSION THREAD

**Discussion by Faltenreich** [LemonAppDev/konsist#1195] [Ask a Question] [answered]
### Wanted end result.

We want to ensure that every Composable is using our AppTheme.

The AppTheme is a function with a lambda parameter containing the child Composables. This method should be called when entering the Compose-world (via `ComposeView.setContent`) to ensure that the AppTheme is applied to every single child Composable. So we need a Konsist test that enforces this structure.

```
@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(content = content)
}
```

The structure of a Composable passing such test could look like this:

```
ComposeView.setContent {
    AppTheme {
        SomeComposable()
    }
}
```

Whereas a Composable failing such test could look like this:

```
ComposeView.setContent {
    // Failure: Not using AppTheme
    SomeComposable()
}
```

Is there any way to achieve this requirement via Konsist, and if not, are we able to use reflection in tandem with Konsist, e.g. to climb up the stack trace in order to ensure a calling AppTheme?

### What you tried so far.

We tried climbing down from `ComposeView.setContent` to ensure that every call of this method directly calls the AppTheme:

```
"ComposeView.setContent calls AppTheme" {
            Konsist.scopeFromProduction()
                .functions()
                // FIXME: Filters only own methods, not inner calls
                .withFullyQualifiedName("androidx.compose.ui.platform.ComposeView.setContent")
                .apply {
                    assertNotEmpty()
                    assertTrue { function ->
                        function.hasLocalFunctionWithName("AppTheme")
                    }
                }
        }
```

We also tried climbing up from every function annotated with `Composable` to ensure that somewhere higher-up the AppTheme has been set.

```
"Composable functions use AppTheme" {
            val composables = Konsist.scopeFromProduction()
                .functions()
                .withAnnotationNamed("Composable")

            composables.assertNotEmpty()

            val appTheme = Konsist.scopeFromProduction()
                .functions()
                .withFullyQualifiedName("de.rewe.app.style.composable.theme.AppTheme")

            appTheme.assertNotEmpty()

            composables.assertTrue { function ->
                // TODO: How to ensure that appTheme has been called before leaving the Compose-world / calling ComposeView.setContent?
            }
        }
```

**Comment by igorwojda** [comment, marked as answer]
Konsist API allows to verify code structure, but ATM it's not possible to query elements in function body.

However every Konsist declaration provides a `text` property to retrieve entire declaration as String. You can use:

```kotlin
assertTrue { function ->
                        function.text.contains("XYZ")
                    }
```
